### ISSUE TITLE
4.1 no longer allows copy and pasting to same VM

### ISSUE THREAD

**Issue by tetrahedras** [QubesOS/qubes-issues#7502]
In 4.0 and earlier, it was possible to copy (Ctrl-Shift-C) text from a given VM and then paste it back to the same VM with Ctrl-Shift-V.

With 4.1, this appears to be no longer possible. Pressing Ctrl-Shift-V on the origin VM no longer results in any action.

Because there are several clipboards in Linux, copying from and then pasting back to that VM was a useful way to work around issues related to these clipboards.

How do I re-enable pasting to the origin VM?

**Comment by tzwcfq** [comment]
Well it'd be actually a handy feature sometimes to have two clipboards available in VM so you can copy two parts of a text from one place in document then scroll to a different part in document and paste text from both clipboards there.

**Comment by tzwcfq** [comment]
If you need to use another VM as a buffer it'll be impossible to use this "two clipboards feature" anymore from the usability point of view as you need to press Alt+Tab and Ctrl+Shift+C/V multiple times as well as from security point of view as you won't be copying your possibly confidential text to secure dom0 anymore but to another VM. And to keep VM only to use as clipboard buffer just doesn't make sense.
Since @tetrahedras said that it worked before then I guess the global clipboard function was reworked and due to the specifics of the implementation or there being some security issues the copy/paste from/to the same VM was unimplemented/disabled.
If it's not a problem from security point and it won't take a lot of resources to implement it then it'd be nice to have this feature again.

**Comment by DemiMarie** [comment]
> Besides, we haven't even established whether this was an intentional change or not. @marmarek, was it?

I believe I was the last person to change this code significantly.  I can make a patch, but it is not trivial, as disposable VM names can be reused.

**Comment by marmarek** [comment]
> as disposable VM names can be reused.

I don't think it's in any way related to the issue discussed here.

In any case, qrexec itself doesn't support loopback connections. Clipboard paste from Linux to Linux is the only operation that uses qrexec policy but underneath isn't really a qrexec connection - that's why it worked previously. But now the policy handling got unified and it prevents such connections regardless of the specific service.

Proper solution is implementing https://github.com/QubesOS/qubes-issues/issues/951.

**Comment by UndeadDevel** [comment]
Just ran into this as well. I want to be able to copy something to global clipboard, restart the VM I copied it from, and then copy it back to that VM (I'm using named DVMs for some of the stuff I do, that's why the newly started VM has the same name). To get more concrete, one of the use cases is that I want the VM to have more CPU cores or RAM, which can only take effect if I restart it. The workaround of copying to another VM is obviously less convenient, especially if I'm copying somewhat sensitive data.

That's why I'm petitioning for #951 to be reopened, as it was closed for EOL reasons, but this issue is still around in 4.1.
